src/java.base/macosx/native/libnio/fs/MacOSXWatchService.c

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, JetBrains s.r.o.. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "jni.h"
+#include "jni_util.h"
+#include "nio_util.h"
+
+#include <CoreFoundation/CoreFoundation.h>
+#include <CoreServices/CoreServices.h>
+
+// MacOSXWatchService.callback()
+static jmethodID        callbackMID;
+// The instance of MacOSXWatchService that is associated with this thread
+static __thread jobject watchService;
+
+JNIEXPORT void JNICALL
+Java_sun_nio_fs_MacOSXWatchService_initIDs(JNIEnv* env, __unused jclass clazz)
+{
+    callbackMID = (*env)->GetMethodID(env, clazz,
+                                      "callback", "(J[Ljava/lang/String;J)V");
+}
+
+extern CFStringRef
+toCFString(JNIEnv *env, jstring javaString);
+
+static jboolean
+convertToJavaStringArray(JNIEnv* env,
+                         char **eventPaths,
+                         const jsize numEventsToReport,
+                         jobjectArray javaEventPathsArray)
+{
+    for (jsize i = 0; i < numEventsToReport; i++) {
+        const jstring path = JNU_NewStringPlatform(env, eventPaths[i]);
+        CHECK_NULL_RETURN(path, FALSE);
+        (*env)->SetObjectArrayElement(env, javaEventPathsArray, i, path);
+    }
+
+    return JNI_TRUE;
+}
+
+static void
+callJavaCallback(JNIEnv* env,
+                 jlong streamRef,
+                 jobjectArray javaEventPathsArray,
+                 jlong eventFlags)
+{
+    if (callbackMID != NULL && watchService != NULL) {
+        // We are called on the run loop thread, so it's OK to use
+        // the thread-local reference to the watch service.
+        (*env)->CallVoidMethod(env, watchService, callbackMID,
+                               streamRef, javaEventPathsArray, eventFlags);
+    }
```

Comment:
Is there a way to notify of events without an upcall? Maybe add the necessary values to a queue of some sort observed by the Java layer?

Reply:
I'm not sure I know how to synchronize native code with Java without any upcall to the Java side; but maybe you are suggesting something as simple as a `notify()` call in place of the current callback with many arguments?.

And also this goes against my understanding that the native part has to be kept as simple and as straightforward as possible. Having it  maintain a queue that has to be synchronized with Java will certainly complicate the native part.

Reply:
I was thinking of maybe something like a [BlockingQueue](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/concurrent/BlockingQueue.html) field on `sun.nio.fs.MacOSXWatchService`, but I am not sure whether this is any more correct.

If the callback is retained, it might be worth considering using [JNU_CallMethodByName](https://github.com/openjdk/jdk/blob/5285035ed9bb43a40108e4d046e0de317730f193/src/java.base/share/native/libjava/jni_util.h#L134).